src/measure_theory/constructions/pi.lean

```diff
@@ -496,6 +500,46 @@ begin
   exacts [ennreal.prod_lt_top (λ i _, hμ i), λ i, (ho i).measurable_set]
 end
 
+variable (μ)
+
+/-- Separating the indices into those that satisfy a predicate `p` and those that don't maps
+a product measure to a product of product measures. This is useful to apply Fubini to some subset
+of the variables. The converse is `measure_theory.measure.map_pi_equiv_pi_subtype_prod`. -/
+lemma map_pi_equiv_pi_subtype_prod_symm (p : ι → Prop) [decidable_pred p] :
+  map (equiv.pi_equiv_pi_subtype_prod p α).symm
+    (measure.prod (measure.pi (λ i, μ i)) (measure.pi (λ i, μ i))) = measure.pi μ :=
+begin
+  refine (measure.pi_eq (λ s hs, _)).symm,
+  have A : (equiv.pi_equiv_pi_subtype_prod p α).symm ⁻¹' (set.pi set.univ (λ (i : ι), s i)) =
+    set.prod (set.pi set.univ (λ i, s i)) (set.pi set.univ (λ i, s i)),
+  { ext x,
+    simp only [equiv.pi_equiv_pi_subtype_prod_symm_apply, mem_prod, mem_univ_pi, mem_preimage,
+      subtype.forall],
+    split,
+    { exact λ h, ⟨λ i hi, by simpa [dif_pos hi] using h i,
+                  λ i hi, by simpa [dif_neg hi] using h i⟩ },
+    { assume h i,
+      by_cases hi : p i,
+      simpa only [dif_pos hi] using h.1 i hi,
+      simpa only [dif_neg hi] using h.2 i hi } },
+  rw [measure.map_apply (measurable_pi_equiv_pi_subtype_prod_symm _ p)
+        (measurable_set.univ_pi_fintype hs), A,
+      measure.prod_prod, pi_pi, pi_pi, ← fintype.prod_subtype_mul_prod_subtype p (λ i, μ i (s i))],
+  { exact λ i, hs i.1 },
+  { exact λ i, hs i.1 },
+  { exact measurable_set.univ_pi_fintype (λ i, hs i.1) },
+  { exact measurable_set.univ_pi_fintype (λ i, hs i.1) },
```

Comment:
Would it be possible to put these simple proofs next to the lemmas they are used for?

Reply:
Unfortunately no. That's what I had tried at first, but Lean doesn't have enough information at the rewrite time to figure out the types, and is confused.